### ISSUE TITLE
[Question]The Demisions update is not sync which is different with native behavior

### ISSUE THREAD

**Issue by NE-SmallTown** [necolas/react-native-web#1662]
At now, when the Demisions update(i.g window `resize`), it will [fire](https://github.com/necolas/react-native-web/blob/df1b62cd454565cc3d07662f84848b483c6f8744/packages/react-native-web/src/exports/Dimensions/index.js#L104) a `change` event. I think maybe this is not the right decision due to 3 reasons:

1. This means we need to add **many** async logic in our APP src code if they relay on the `Demisions`. It will change our code flow(like `async/await`), but is unnessary because it's just a `get` API. Imagine call `window.location` or `window.getElementById` is a async operation, and then how difference it will be to your code. There is not a async system like redux or koa and should not be for Demisions , so make this async is more harm than good.

2. I guess the reason we add the `addEventListener` to Demisions is for the performance issue, otherwise it will trigger the forcible browser `layout`/`paint`(Chrome) or `reflow`(Firefox). But we have added the [cache](https://github.com/necolas/react-native-web/blob/df1b62cd454565cc3d07662f84848b483c6f8744/packages/react-native-web/src/exports/Dimensions/index.js#L43) to `Demisions` object, and the `resize` operation is a low probability event, there won't be frequent `resize` fired even it's a `iframe` rather than a normal `window`. So, for the infrequent case to do many changes described above is not appropriate from my personal view. If I really need that(an async or `debounce`), I can add it manually.

3. The other important thing is, on the native side, it will return the 'right' Demisions **synchronously**.

Therefore, I think maybe change the current behaviro would be better :)

**Comment by necolas** [comment]
I dont understand what you're asking. Dimensions isnt async

**Comment by NE-SmallTown** [comment]
@necolas  Sorry for not clear. You can just put below code in your Chrome Devtools Console:

```javascript
function debounce(func, wait, immediate){
  var timeout, args, context, timestamp, result;
  if (null == wait) wait = 100;

  function later() {
    var last = Date.now() - timestamp;

    if (last < wait && last >= 0) {
      timeout = setTimeout(later, wait - last);
    } else {
      timeout = null;
      if (!immediate) {
        result = func.apply(context, args);
        context = args = null;
      }
    }
  };

  var debounced = function(){
    context = this;
    args = arguments;
    timestamp = Date.now();
    var callNow = immediate && !timeout;
    if (!timeout) timeout = setTimeout(later, wait);
    if (callNow) {
      result = func.apply(context, args);
      context = args = null;
    }

    return result;
  };

  debounced.clear = function() {
    if (timeout) {
      clearTimeout(timeout);
      timeout = null;
    }
  };
  
  debounced.flush = function() {
    if (timeout) {
      result = func.apply(context, args);
      context = args = null;
      
      clearTimeout(timeout);
      timeout = null;
    }
  };

  return debounced;
}

const canUseDOM = true
const win = canUseDOM
  ? window
  : {
      devicePixelRatio: undefined,
      innerHeight: undefined,
      innerWidth: undefined,
      screen: {
        height: undefined,
        width: undefined
      }
    };

const dimensions = {};
const listeners = {};

class Dimensions {
  static get(dimension) {
    return dimensions[dimension];
  }

  static _update() {
    console.log('Dimensions _update!!!')
    dimensions.window = {
      fontScale: 1,
      height: win.innerHeight,
      scale: win.devicePixelRatio || 1,
      width: win.innerWidth
    };

    dimensions.screen = {
      fontScale: 1,
      height: win.screen.height,
      scale: win.devicePixelRatio || 1,
      width: win.screen.width
    };

    if (Array.isArray(listeners['change'])) {
      listeners['change'].forEach(handler => handler(dimensions));
    }
  }

  static addEventListener(
    type,
    handler
  ) {
    listeners[type] = listeners[type] || [];
    listeners[type].push(handler);
  }

  static removeEventListener(
    type,
    handler
  ) {
    if (Array.isArray(listeners[type])) {
      listeners[type] = listeners[type].filter(_handler => _handler !== handler);
    }
  }
}

Dimensions._update();

if (canUseDOM) {
  window.addEventListener('resize', debounce(Dimensions._update, 16), false);
}

window.addEventListener(
  'resize', 
  () => {
    console.log('My resize listener fired!!!');
    console.log(`Dimensions.get('window').width: `, Dimensions.get('window').width);
    console.log('Directly call window.innerWidth: ', window.innerWidth);
        
    setTimeout(() => {
      console.log(`setTimeout Dimensions.get('window').width: `, Dimensions.get('window').width);
      console.log('setTimeout Directly call window.innerWidth: ', window.innerWidth);
     }, 100);
  }, 
  false
);
```

First, open Chrome Devtools, select the 'undock into seperate window' mode(otherwise it will fire `resize` event twice due to the changing of Devtools itself).

![image](https://user-images.githubusercontent.com/18418010/86870606-48e52800-c10b-11ea-8a97-9af520e3c676.png)

And then, click the 'resize button' of the browser:

![image](https://user-images.githubusercontent.com/18418010/86870982-fa845900-c10b-11ea-91d0-b2f259a16168.png)

or any operation could trigger the `resize` event fired, like adjust the browser window size.

And then, the console will output:

![image](https://user-images.githubusercontent.com/18418010/86871462-eab94480-c10c-11ea-9320-b458c2651f71.png)


This is why I file this issue, and the reson had been described above. Sorry for the unclear again.

**Comment by necolas** [comment]
Oh the denounce. Yeah we can probably remove that and make it a user space decision

**Comment by yairopro** [comment]
Thanks @necolas 
Removing the debounce from the `Dimensions` module was a great idea as it now let us choose to debounce, or not, the stream of updates (like you said).

I would still propose to debounce the updates from the `useWindowDimensions` hook since it triggers a render of the components using it at each update.

What do you think @necolas ?

**Comment by necolas** [comment]
Yeah maybe. I wonder why this (probably?) isn't an issue for RN...

**Comment by yairopro** [comment]
>  I wonder why this (probably?) isn't an issue for RN

I think it's because resizing a window on mobile isn't direct as it is in desktop. The system let the user preview the future window size, and resize it only once (when the user release). While on desktop, a browser window is directly resized by the user in real-time.

https://user-images.githubusercontent.com/8110568/146676211-7f68fec1-dc26-40fc-a367-6c51829b2ab1.mp4

**Comment by necolas** [comment]
That's a good point, but Meta uses RN on Desktop to build apps like Messenger Desktop too. That could have the same problem. Please could you create a new issue about this and I'll follow up with the RN Desktop team after the holidays to see if there's a shared problem. Thank you!
